File: java/src/main/java/com/fixcraft/basefwx/X25519.java
/*
 * BaseFWX - Cryptography Engine
 * Copyright (C) 2020-2026  FixCraft Inc.
 * Licensed under the GNU Lesser General Public License v3.0 or later.
 */

package com.fixcraft.basefwx;

import org.bouncycastle.crypto.params.X25519PrivateKeyParameters;
import org.bouncycastle.crypto.params.X25519PublicKeyParameters;

import java.security.SecureRandom;
import java.util.Arrays;

/**
 * X25519 building block mirroring {@code basefwx::x25519} (raw 32-byte keys,
 * wipe private material, reject the all-zero shared secret). Not wired into
 * fwxAES / keywrap file formats.
 *
 * <p>Android sync-list note: when the Android Yume tree is next updated, add
 * {@code X25519.java} to {@code syncBasefwxJava} (Rule 12).
 */
public final class X25519 {
    public static final int KEY_LEN = 32;

    private X25519() {}

    public static final class KeyPairResult {
        public final byte[] publicKey;
        public final byte[] privateKey;

        public KeyPairResult(byte[] publicKey, byte[] privateKey) {
            this.publicKey = publicKey;
            this.privateKey = privateKey;
        }

        /** Wipe the private scalar. Call when done with the keypair. */
        public void wipePrivate() {
            if (privateKey != null) {
                Arrays.fill(privateKey, (byte) 0);
            }
        }
    }

    public static KeyPairResult generateKeyPair() {
        X25519PrivateKeyParameters privateKey =
                new X25519PrivateKeyParameters(new SecureRandom());
        return new KeyPairResult(
                privateKey.generatePublicKey().getEncoded(),
                privateKey.getEncoded());
    }

    public static byte[] deriveSharedSecret(byte[] privateKey, byte[] peerPublicKey) {
        if (privateKey == null || privateKey.length != KEY_LEN) {
            throw new IllegalArgumentException("X25519 private key must be 32 bytes");
        }
        if (peerPublicKey == null || peerPublicKey.length != KEY_LEN) {
            throw new IllegalArgumentException("X25519 public key must be 32 bytes");
        }
        X25519PrivateKeyParameters local =
                new X25519PrivateKeyParameters(privateKey);
        X25519PublicKeyParameters peer =
                new X25519PublicKeyParameters(peerPublicKey);
        byte[] shared = new byte[KEY_LEN];
        try {
            local.generateSecret(peer, shared, 0);
        } catch (IllegalStateException exc) {
            Arrays.fill(shared, (byte) 0);
            throw new IllegalArgumentException(
                    "X25519 peer produced the forbidden all-zero shared secret", exc);
        }
        if (isAllZero(shared)) {
            Arrays.fill(shared, (byte) 0);
            throw new IllegalArgumentException(
                    "X25519 peer produced the forbidden all-zero shared secret");
        }
        return shared;
    }

    private static boolean isAllZero(byte[] data) {
        for (byte b : data) {
            if (b != 0) {
                return false;
            }
        }
        return true;
    }
}

File: python/basefwx/crypto/_x25519.py
# BaseFWX - Cryptography Engine
# Copyright (C) 2020-2026  FixCraft Inc.
# Licensed under the GNU Lesser General Public License v3.0 or later.

"""X25519 building block mirroring ``basefwx::x25519`` / Java ``X25519``.

Raw 32-byte keys, wipe helpers, reject the all-zero shared secret.
Not wired into fwxAES / keywrap file formats.
"""

from __future__ import annotations

from cryptography.hazmat.primitives.asymmetric.x25519 import (
    X25519PrivateKey,
    X25519PublicKey,
)


KEY_LEN = 32


class KeyPair:
    __slots__ = ("public_key", "private_key")

    def __init__(
        self, public_key: bytes, private_key: "bytes | bytearray"
    ) -> None:
        self.public_key = public_key
        self.private_key = bytearray(private_key)

    def wipe_private(self) -> None:
        self.private_key[:] = b"\x00" * len(self.private_key)

    def __del__(self) -> None:  # pragma: no cover - best-effort wipe
        try:
            self.wipe_private()
        except Exception:
            pass


def generate_keypair() -> KeyPair:
    private = X25519PrivateKey.generate()
    public = private.public_key()
    return KeyPair(
        public_key=public.public_bytes_raw(),
        private_key=private.private_bytes_raw(),
    )


def derive_shared_secret(
    private_key: "bytes | bytearray | memoryview",
    peer_public_key: "bytes | bytearray | memoryview",
) -> bytes:
    if len(private_key) != KEY_LEN:
        raise ValueError("X25519 private key must be 32 bytes")
    if len(peer_public_key) != KEY_LEN:
        raise ValueError("X25519 public key must be 32 bytes")
    local = X25519PrivateKey.from_private_bytes(bytes(private_key))
    peer = X25519PublicKey.from_public_bytes(bytes(peer_public_key))
    shared = local.exchange(peer)
    if len(shared) != KEY_LEN:
        raise RuntimeError("X25519 shared-secret size mismatch")
    if shared == b"\x00" * KEY_LEN:
        raise ValueError("X25519 peer produced the forbidden all-zero shared secret")
    return shared
